main_test.go

```diff
@@ -15,6 +16,28 @@ func TestMain(t *testing.T) {
 	})
 }
 
+// TestMissingCommand will cause the main function to return a non-zero exit code
+// so we need to do some wrapping to make the test not fail.
+// For more info check ou to uset: https://talks.golang.org/2014/testing.slide#23
+func TestMissingCommand(t *testing.T) {
```

Comment:
This isn't a "Unit test" I think it may be moved to `tests/ergo_test.go`

Reply:
The way this is tested is different of the way the rest of the command are tested there.
This would be an integration test. I say this because it has to do with interacting with an external component (**os.args**), but otherwise is an unit test.
 
To make it a simple unit test it would suffice to add an init() method that would set an array to os.args. Then you could use that array in the program, and use that array in the tests.

The way it tested is the same the ```exec.Command``` is tested in golang itself.

The other tests (in ```ergo_test.go``` that is) would be end to end or acceptance tests.

For this one you do not need a compiled executable and, further more it is seen by ```-cover```.
The ones in ```ergo_test.go``` are based on a compiled binary and would never be seen by ```-cover```.

This should not be merged with ```ergo_test.go``` in my opinion, perhaps a further test could be implemented there. The name of the make target should also be changed to ```e2e``` perhaps for ```ergo_test.go```.

Reply:
This is the same as all of the other tests in this file except for one detail, it's dealing with a non-zero exit code. When we do an `os.Exit(1)` inside of a command, `go test` will panic. Wrapping our test call allows it to catch the panic from `go test`. There isn't an easy way to do this otherwise and this format of test is used to test the go toolset as well as provided in the comment. This test though still does just test one function using go test and does not test an actual binary so I would keep it here.

Reply:
My only concern with this test is that it rely on `exec.Command` which spins a different process and If you noticed it is beeing created with `os.Args[0]` as the command name which is `ergo` so it is spinning and ergo process and is getting the result of it. We already had some problems with tests like these in the past and that why we created the "Integration tests" (all in`tests/ergo_test.go`) for tests that rely on such external resource. If you have no `ergo` binary in your system it is gonna fail.
```
...
cmd := exec.Command(os.Args[0], "-test.run=TestMissingCommand") 
cmd.Env = append(os.Environ(), "TEST_MISSING_COMMAND=1")
err := cmd.Run()
...
```

Plus it isn't reaching the lines 27~33. I tried by adding some logs.

I know why you did this is because you were having the error `flag redefined: h` so you are spining another process just to avoid this error. 

Please take a look in this commit I think it could be a solution for this case.
https://github.com/cristianoliveira/ergo/commit/3eac4ce51060452f36cfe12d0a0b123b0791f2ae

Reply:
So what ends up happening:

1. Start this test run.
2. Check the environment variable "TEST_MISSING_COMMAND" and see if it is set to the string "1". This is run by the exec command.
    a. If it is, then change the arguments to the command to be `ergo`. 
    b. Call the main function which will exit with an exit code of 1 which will cause the go test command executed below to panic. This will cause this test to fail, but it is in a separate **test** command process. The if statement will stop executing here since there was a panic and the calling process will take over.
3. In the original process, we will take the **original** executable that started the original test run and create a command that will only run this test again. This is set to a test program built by `go test ...` and **not** a compiled version of `ergo`. You can add a log statement such as this to see what the original `os.Args` is set to: `tt.Fatal(os.Args)`
4. We set an environment variable for the run of that command to make sure that our if statement in 2 above will be exected and thus, run our unit test
5. Run the command..
6. Check to make sure that the program exited with an exit code of 1 to see that it successfully failed and return.

Reply:
Okay I got everything you were doing in this code. The thing you are missing is that you are currently making a `ergo` process when you do `exec.Command(os.Args[0], "-test.run=TestMissingCommand")` you original intention was run a separate test didn't you?

But check what is in `os.Args[0]` and you will see. Is the same as `exec.Command('ergo', "-test.run=TestMissingCommand")`. Which only works because you have an actually binary ergo in your `PATH`. It may not be true when running the normal tests `make test`.

It only ensured when you run it using `make test-integration` because it builds it before

Reply:
So My problem was that I was running only the test I cared about when I was checking to make sure that it was working correctly and forgot that the rest of the tests are mutating global state that affect all tests. I've updated the test to keep track of the original test executable, which is just for the tests, and uses that in the `exec` call.

Reply:
No worries :)